src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -224,24 +221,28 @@ void DBusService::userRemoved(uint uid, const QDBusObjectPath& path) {
   m_appTracker->userRemoved(uid);
 }
 
-void DBusService::appLaunched(const QString& cgroup, const QString& appId,
-                              int rootpid) {
-  logger.debug() << "tracking:" << cgroup << "appId:" << appId
-                 << "PID:" << rootpid;
+void DBusService::appLaunched(const QString& cgroup, const QString& desktopId) {
+  logger.debug() << "tracking:" << cgroup << "id:" << desktopId;
+
+  AppState state = m_excludedApps.value(desktopId, Active);
+  if (state == Active) {
+    // Nothing to do here.
+    return;
+  }
 
-  // HACK: Quick and dirty split tunnelling.
-  // TODO: Apply filtering to currently-running apps too.
-  if (m_excludedApps.contains(appId)) {
+  // Apply firewall rules to this control group.
+  m_excludedCgroups[cgroup] = state;
+  if (state == Excluded) {
     m_wgutils->excludeCgroup(cgroup);
   }
 }
 
-void DBusService::appTerminated(const QString& cgroup, const QString& appId) {
-  logger.debug() << "terminate:" << cgroup;
+void DBusService::appTerminated(const QString& cgroup,
+                                const QString& desktopId) {
+  logger.debug() << "terminate:" << cgroup << "id:" << desktopId;
 
-  // HACK: Quick and dirty split tunnelling.
-  // TODO: Apply filtering to currently-running apps too.
-  if (m_excludedApps.contains(appId)) {
+  // Remove any firewall rules applied to this control group.
```

Comment:
Shouldn't this app also be removed from m_excludedApps?

Reply:
No, the gist of this, is that m_excludedApps holds the applications that the user wishes to exclude, so it kind of hangs around as a persistent setting, but m_excludedCgroups holds the matched applications that are currently running.

This signal occurs when the application closes, in which case the cgroup has been destroyed, but we still need to remember the setting in case the user re-opens it.

Reply:
Is there any risk of desktopId re-use that can impact m_excludedApps, because the ID is not removed when the app is terminated? For example, could an excluded app be uinstalled and another app installed with the same desktopId while the VPN is on? Or is this an unlikely edge case that is difficult to handle?

Reply:
I think this scenario would be commonly encountered if a user updates software on their device while the VPN is active, and I think the correct behavior in that case is to treat the newly-installed application identically to the previous version.

The recommendation in the freedesktop specification is for application names to follow a reverse-DNS naming convention, in which case they are expected to be a kind of globally unique identifier.